link.go

```diff
@@ -94,18 +93,16 @@ func (l *link) attach(ctx context.Context, beforeAttach func(*frames.PerformAtta
 	var fr frames.FrameBody
 	select {
 	case <-ctx.Done():
-		// attach was written to the network. assume it was received
-		// and that the ctx was too short to wait for the ack.
-		go func() {
-			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
-			defer cancel()
-			l.muxClose(ctx, nil, nil, nil)
-		}()
-		return ctx.Err()
+		// we are now in an inconsistent state, so kill the connection
```

Comment:
This might not work for Service Bus when using sessions.

SB lets you ask for the "next available session", which can take some time if there aren't any available sessions. The typical use case would be iterating, continually calling SBClient.AcceptNextSession(), which in turn attempts to create an amqp.Receiver with a ReceiverOptions.Filters. When there are no available sessions the AcceptNextSession(ctx) call will block, and the service will eventually attach a receiver or send back a timeout error, which we parse and return.

I think it's quite possible that a user might cancel when they reach the end - ie, they don't really want to wait the full "1 minute + retries" that we do. If that happens we'll end up killing the entire connection, which would then invalidate all their other SB Session receivers.

Perhaps we could make this a LinkOption where the cancellation signal will simply clean up our local state - if the remote service did end up telling us to attach a link it'd just fail.

Reply:
I'm unclear on the behavior here.  Are you saying that the underlying call to `amqp.Session.NewReceiver()` will block if there's no session available?

Reply:
Also, similar to the above issue with sessions, even if we were to clean up local state and keep going, if we later receive the attach ack for an unknown link, this will kill the connection.